File: mprt/python/mprt.py
#! /usr/bin/env python3.6
# rosalind.info/mprt

import re
import sys
import requests


def read_file(file_path):
    return open(file_path, "r").read().splitlines()


def get_fasta_from_id(uniprot_id):
    url = f"http://www.uniprot.org/uniprot/{uniprot_id}.fasta"
    fasta = requests.get(url).text.split("\n")[1:] # remove first line
    return "".join([line.strip() for line in fasta])


def find_motif(seq, motif="N[^P][ST][^P]", overlap=True):
    if overlap: # use a lookahead
        motif = f"(?={motif})"
    return [position.start()+1 for position in re.finditer(motif, seq)]


def main():
    uniprot_ids = read_file(sys.argv[1])
    for uniprot_id in uniprot_ids:
        fasta = get_fasta_from_id(uniprot_id)
        motifs = find_motif(fasta)
        if len(motifs) > 0:
            print(uniprot_id)
            print(*motifs)


if __name__ == "__main__":
    main()

File: mprt/nim/mprt.nim
# rosalind.info/mprt

import httpclient, strutils, re, sequtils, os

proc read_file(file_path: string): seq[string] =
  # read in file, return sequence, an element per line
  result = @[]
  let file = open(file_path)
  for line in file.lines:
    result.add(line)


proc get_fasta(uniprot_id: string): string =
  # get fasta sequence from uniprot
  var
    url = "http://www.uniprot.org/uniprot/$#.fasta" % uniprot_id
    text = split_lines(newHttpClient().get_content(url))
  text.delete(0)
  return foldl(text, a&b)


proc find_motif(sequence: string,
                motif: Regex = re"(?=N[^P][ST][^P])"): seq[int] =
  # find motif positions in sequence using regex and lookaheads
  result = @[]
  var last_position = 0
  while last_position != -1:
    last_position = find(sequence, motif, start=last_position+1)
    if last_position != -1:
      result.add(last_position)
    # add 1 to positions (zero-indexed => 1-indexed)
  result = result.map(proc(x:int): int= x+1)


proc print(x: seq, sep: string = " ") =
  #print sequence with a separator
  for i in x:
    stdout.write i, sep


for motif in read_file(param_str(1)):
  var
    fasta = get_fasta(motif)
    output = find_motif(fasta)
  if len(output) != 0:
    echo motif
    print(output)
